[inductor] decompose memory bound mm

jackiexu1992 opened the pull request:
Summary:
Decompose memory bound mm/bmm. 
Linear decomposition result:  D53502768
BMM decomposition result: D53148650
 We should only decompose when
1)bmm, b is large, m,n,k is relative small
2)mm/addmm. m is large, n and K is relative small. e.g. mm of input gradient in linear backward should not be decomposed since m is small and n is large.
Need to conduct more experiments to see if we can find a better strategy for decomposition. I have tried to use a linear regression model (see the bento results) which does not fit well. For short term, we use heuristics to determine when to decompose.

Test Plan:
```
buck2 test mode/dev-nosan //caffe2/test/inductor:decompose_mem_bound_mm
```

COFFEE APS mc0: 
baseline: aps-lsf-0124-bf16-267ccb7a0d
decompose: aps-lsf-0124-bf16-4e3824db40

FIRST AFOC pyper mc1

Differential Revision: D53602514




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

eellison reviewed (commented):
Can you post benchmarks varying M, N, and K ? Would you also submit a PT2 dashboard benchmark run https://github.com/pytorch/pytorch/actions/workflows/inductor-perf-test-nightly.yml ?

eellison commented on torch/_inductor/fx_passes/decompose_mem_bound_mm.py:
```diff
@@ -0,0 +1,123 @@
+    return a.is_cuda and b.is_cuda
+
+
+def should_decompose_common(
```
Can we check that the matmul is actual memory bound here ? 

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/pad_mm.py#L180

jackiexu1992 replied:
![image](https://github.com/pytorch/pytorch/assets/10678066/4cb85cd2-514c-4452-88de-a1644da9d4b0)
![image](https://github.com/pytorch/pytorch/assets/10678066/77207cda-0b88-4e54-862a-c2cd648fbe01)

I have tried to use 2 ways to calculate compute intensity. 1/(1/m+1/n+1/k), m*n*k/(m*n+n*k+m*k). However, the result shows that not all the memory bound mms can be decomposed.

eellison commented:
@jackiexu1992  adding the label doesnt do anything. you need to manually trigger a run from the page i linked from this branch. ive only had luck selecting from a ghstack branch

eellison reviewed (commented):
@Chillee found that we could only beat CuBlas for certain memory bound matrix multiplications when coordinate_descent_autotune is set. 

cc @Chillee care to comment ?

jackiexu1992 commented:
> @jackiexu1992 adding the label doesnt do anything. you need to manually trigger a run from the page i linked from this branch. ive only had luck selecting from a ghstack branch

how to trigger a run from this branch? searched by branch name jackiexu1992:export-D53602514. Did not find it.

eellison commented:
Can you submit this as a ghstack and then choose it ?

Chillee commented:
> cc @Chillee care to comment ?

Yeah, just that Inductor generated gemv's don't consistently beat cublas unless we do coordinate descent tuning lol

jackiexu77 committed:
[inductor] decompose memory bound mm (#120047)

Summary:

Decompose memory bound mm/bmm. 
Linear decomposition result:  D53502768
BMM decomposition result: D53148650
 We should only decompose when
1)bmm, b is large, m,n,k is relative small
2)mm/addmm. m is large, n and K is relative small. e.g. mm of input gradient in linear backward should not be decomposed since m is small and n is large.
Need to conduct more experiments to see if we can find a better strategy for decomposition. I have tried to use a linear regression model (see the bento results N4955775) which does not fit well. For short term, we use heuristics to determine when to decompose.

Compute intensity vs speedup gain

 {F1459185346}

Test Plan:
```
buck2 test mode/dev-nosan //caffe2/test/inductor:decompose_mem_bound_mm
```
FIRST AFOC pyper mc1  (SimpleLCE)
pt2 decompose: f533788703
optimize lce: f533730079
baseline: f533746677

~19% qps gain for first model
 {F1459472526} 
The results show that pt2 decomposition can keep qps neutral with ne unchanged. 

2k event LCE:
Test on local model.
decompose_bmm: average latency is 496.8038330078125 ms.
https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/test/jackiexu0313/jackie_test/devgpu005.cln5.facebook.com/pid23302911708457758_random6697323698.json.gz&bucket=perfdoctor_gpu_traces_test 

decompose pt2: average latency is 493.80352783203125 ms.
https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/test/jackiexu0313/jackie_test/devgpu005.cln5.facebook.com/pid37668601708460273_random2028909216.json.gz&bucket=perfdoctor_gpu_traces_test 

QPS on par. 


COFFEE APS mc0: (LCE)
baseline: aps-lsf-0124-bf16-267ccb7a0d
decompose: aps-lsf-0124-bf16-4e3824db40
blocked by error. https://fb.workplace.com/groups/1075192433118967/permalink/1381918665779674/

Reviewed By: Chillee

Differential Revision: D53602514
